### PR TITLE
Unit tests for check security rules.

### PR DESCRIPTION
**Pull request by sarvaniv**

### CODE REVIEW

File: contrib/endpoints/src/api_manager/check_security_rules_test.cc [hunk 1]
```diff
@@ -0,0 +1,413 @@
...
+        std::string body;
+        req->OnComplete(Status(Code::INTERNAL, "Cannot talk to server"),
+                        std::move(empty), std::move(body));
+      }));
+
+  CheckSecurityRules(request_context_,
+                     [](Status status) { ASSERT_FALSE(status.ok()); });
+}
+
+TEST_F(CheckSecurityRulesTest, CheckAutzFailWithTestResultFailure) {
```
Comment (liminw) [inline contrib/endpoints/src/api_manager/check_security_rules_test.cc]: Authz

File: contrib/endpoints/src/api_manager/check_security_rules_test.cc [hunk 2]
```diff
@@ -0,0 +1,413 @@
...
+
+  MockApiManagerEnvironment *raw_env_;
+  MockRequest *raw_request_;
+  std::shared_ptr<context::RequestContext> request_context_;
+  std::shared_ptr<context::ServiceContext> service_context_;
+  std::string release_url_;
+  std::string ruleset_test_url_;
+};
+
+TEST_F(CheckSecurityRulesTest, CheckAuthzFailGetRelease) {
```
Comment (liminw) [inline contrib/endpoints/src/api_manager/check_security_rules_test.cc]: Can you add comments for each test, explaining what the test setup and expectation?

File: contrib/endpoints/src/api_manager/check_security_rules_test.cc [hunk 3]
```diff
@@ -0,0 +1,413 @@
...
+        new ::testing::NiceMock<MockRequest>());
+    raw_request = request.get();
+
+    request_context_ = std::make_shared<context::RequestContext>(
+        service_context_, std::move(request));
+    EXPECT_CALL(*raw_env_, DoRunHTTPRequest(_)).Times(0);
+  }
+
+  MockApiManagerEnvironment *raw_env_;
+  MockRequest *raw_request;
```
Comment (qiwzhang) [inline contrib/endpoints/src/api_manager/check_security_rules_test.cc]: do you need to store two raw pointers in the class?

File: contrib/endpoints/src/api_manager/check_security_rules_test.cc [hunk 4]
```diff
@@ -0,0 +1,413 @@
...
+
+  EXPECT_CALL(*raw_env_,
+              DoRunHTTPRequest(
+                  AllOf(Property(&HTTPRequest::url, StrEq(ruleset_test_url_)),
+                        Property(&HTTPRequest::method, StrCaseEq("POST")))))
+      .Times(0);
+
+  auto ptr = this;
+  CheckSecurityRules(request_context_, [ptr](Status status) {
+    ptr->raw_env_->LogInfo(status.ToString());
```
Comment (qiwzhang) [inline contrib/endpoints/src/api_manager/check_security_rules_test.cc]: can we verify the error string from status?  Log will not help

File: contrib/endpoints/src/api_manager/check_security_rules_test.cc [hunk 5]
```diff
@@ -0,0 +1,413 @@
...
+
+TEST_F(CheckSecurityRulesTest, CheckAuthzFailGetRelease) {
+  std::string service_config = std::string(kBadServiceName) +
+                               kProducerProjectId + kApis + kAuthentication +
+                               kHttp;
+  std::string server_config = kServerConfig;
+  SetUp(service_config, server_config);
+
+  EXPECT_CALL(*raw_env_, DoRunHTTPRequest(AllOf(
+                             Property(&HTTPRequest::url, StrNe(release_url_)),
```
Comment (qiwzhang) [inline contrib/endpoints/src/api_manager/check_security_rules_test.cc]: shouldn't it be StrEQ?

File: contrib/endpoints/src/api_manager/check_security_rules_test.cc [hunk 6]
```diff
@@ -0,0 +1,413 @@
...
+  CheckSecurityRules(request_context_, [](Status status) {
+    ASSERT_TRUE(status.CanonicalCode() == Code::PERMISSION_DENIED);
+  });
+}
+
+TEST_F(CheckSecurityRulesTest, CheckAuthzSuccess) {
+  std::string service_config = std::string(kServiceName) + kProducerProjectId +
+                               kApis + kAuthentication + kHttp;
+  std::string server_config = kServerConfig;
+  SetUp(service_config, server_config);
```
Comment (qiwzhang) [inline contrib/endpoints/src/api_manager/check_security_rules_test.cc]: It seems all your tests are using the same Setup.  Can we use the ::Test::SetUp()?

File: contrib/endpoints/src/api_manager/check_security_rules_test.cc [hunk 7]
```diff
@@ -0,0 +1,413 @@
...
+  });
+}
+
+TEST_F(CheckSecurityRulesTest, CheckAuthzFailTestRuleset) {
+  std::string service_config = std::string(kServiceName) + kProducerProjectId +
+                               kApis + kAuthentication + kHttp;
+  std::string server_config = kServerConfig;
+  SetUp(service_config, server_config);
+
+  request_context_->set_auth_claims(kJwtEmailPayload);
```
Comment (qiwzhang) [inline contrib/endpoints/src/api_manager/check_security_rules_test.cc]: Do we have a test case to test no auth_claims

Comment (sarvaniv) [inline contrib/endpoints/src/api_manager/check_security_rules_test.cc]: Yes. This is when the auth is disabled. The test cases are added/

File: contrib/endpoints/src/api_manager/check_security_rules_test.cc (hunk 5 above)
Comment (sarvaniv) [inline contrib/endpoints/src/api_manager/check_security_rules_test.cc]: This is the fail test case where we have a bad URL. If the release url is not equal to the expected URL then the response is failure.

File: contrib/endpoints/src/api_manager/check_security_rules_test.cc (hunk 4 above)
Comment (sarvaniv) [inline contrib/endpoints/src/api_manager/check_security_rules_test.cc]: This was a debug log that I forgot to remove.

File: contrib/endpoints/src/api_manager/check_security_rules_test.cc (hunk 6 above)
Comment (sarvaniv) [inline contrib/endpoints/src/api_manager/check_security_rules_test.cc]: The SetUp method is part of the test class. But I needed a way to pass configuration to different tests.

File: contrib/endpoints/src/api_manager/check_security_rules_test.cc [hunk 8]
```diff
@@ -0,0 +1,413 @@
...
+namespace {
+const char kServiceName[] = R"(
+name: "myfirebaseapp.appspot.com")";
+
+const char kBadServiceName[] = R"(
+name: "badService.appspot.com")";
+
+const char kProducerProjectId[] = R"(
+producer_project_id: "myfirebaseapp"
+)";
```
Comment (wattli) [inline contrib/endpoints/src/api_manager/check_security_rules_test.cc]: Can we make it consist with the two cases above by merging this line to the end of previous line?

File: contrib/endpoints/src/api_manager/check_security_rules_test.cc [hunk 9]
```diff
@@ -0,0 +1,413 @@
...
+    selector: "ListShelves"
+    get: "/ListShelves"
+  }
+}
+control {
+  environment: "http://127.0.0.1:8081"
+})";
+
+const char kJwtEmailPayload[] =
+    R"({"iss":"https://accounts.google.com","iat":1486575396,"exp":1486578996,"aud":"https://myfirebaseapp.appspot.com","sub":"113424383671131376652","email_verified":true,"azp":"limin-429@appspot.gserviceaccount.com","email":"limin-429@appspot.gserviceaccount.com"})";
```
Comment (wattli) [inline contrib/endpoints/src/api_manager/check_security_rules_test.cc]: Same here, can we make this consistent with everything else in this file by merging this line to the end of previous one?

File: contrib/endpoints/src/api_manager/check_security_rules_test.cc [hunk 10]
```diff
@@ -0,0 +1,413 @@
...
+  ]
+}
+)";
+
+std::pair<std::string, std::string> GetConfigWithAuthForceDisabled() {
+  std::string service_config =
+      std::string(kServiceName) + kApis + kAuthentication + kHttp;
+  const char server_config[] = R"(
+api_authentication_config {
+force_disable:  true
```
Comment (wattli) [inline contrib/endpoints/src/api_manager/check_security_rules_test.cc]: indent 2 spaces.

File: contrib/endpoints/src/api_manager/check_security_rules_test.cc (hunk 6 above)
Comment (qiwzhang) [inline contrib/endpoints/src/api_manager/check_security_rules_test.cc]: It seems to me that (service_config, server_config) for all 4 test cases are the same.   If so, you can use test::SetUp().

Comment (qiwzhang) [inline contrib/endpoints/src/api_manager/check_security_rules_test.cc]: Never mind.  I see that the first test case is using badServieName. is different with others

File: contrib/endpoints/src/api_manager/check_security_rules_test.cc [hunk 11]
```diff
@@ -0,0 +1,461 @@
...
+
+// Get service configuration with no authentication member field. This will
+// disable auth and will also disable security rules check.
+std::pair<std::string, std::string> GetConfigWithNoAuth() {
+  std::string service_config = std::string(kServiceName) + kApis + kHttp;
+  return std::make_pair(service_config, std::string(kServerConfig));
+}
+
+// Get Service configuration with no apis. This will result in the version field
+// no present and should disable security rules check.
```
Comment (liminw) [inline contrib/endpoints/src/api_manager/check_security_rules_test.cc]: no present -> not present

File: contrib/endpoints/src/api_manager/check_security_rules_test.cc [hunk 12]
```diff
@@ -0,0 +1,461 @@
...
+  std::shared_ptr<context::RequestContext> request_context_;
+  std::shared_ptr<context::ServiceContext> service_context_;
+  std::string release_url_;
+  std::string ruleset_test_url_;
+};
+
+// If the release name is bad, then check the following:
+// 1. Ensure that GetRuleset request is inovked on bad release name.
+// 2. In this case return Status with NOT_FOUND
+// 3. Ensure that there are no mor HTTP calls made to firbase TestRuleset
```
Comment (liminw) [inline contrib/endpoints/src/api_manager/check_security_rules_test.cc]: typo: more

**Pull request merged**